pkg/server/backend_manager.go

```diff
@@ -214,42 +235,40 @@ func containIDType(idTypes []header.IdentifierType, idType header.IdentifierType
 }
 
 // AddBackend adds a backend.
-func (s *DefaultBackendStorage) AddBackend(identifier string, idType header.IdentifierType, conn agent.AgentService_ConnectServer) Backend {
+func (s *DefaultBackendStorage) AddBackend(identifier string, idType header.IdentifierType, backend Backend) {
 	if !containIDType(s.idTypes, idType) {
 		klog.V(4).InfoS("fail to add backend", "backend", identifier, "error", &ErrWrongIDType{idType, s.idTypes})
-		return nil
+		return
 	}
-	klog.V(5).InfoS("Register backend for agent", "connection", conn, "agentID", identifier)
+	klog.V(5).InfoS("Register backend for agent", "connection", backend, "agentID", identifier)
 	s.mu.Lock()
 	defer s.mu.Unlock()
 	_, ok := s.backends[identifier]
-	addedBackend := newBackend(conn)
 	if ok {
 		for _, v := range s.backends[identifier] {
-			if v.conn == conn {
-				klog.V(1).InfoS("This should not happen. Adding existing backend for agent", "connection", conn, "agentID", identifier)
-				return v
+			if v == backend {
+				klog.V(1).InfoS("This should not happen. Adding existing backend for agent", "connection", backend, "agentID", identifier)
```

Comment:
I doubt `backend` will serialize well, maybe just drop it from this log? Also consider adding ID type.

Also, if this shouldn't happen, should it be error level?

```suggestion
				klog.V(1).ErrorS(nil, "Adding duplicate backend for agent. This should not happen", "agentID", identifier, "idType", idType)
```

Reply:
I had just removed it (in order to fix the test race) as you sent this feedback. :)

I don't understand the "This should not happen" codepaths in depth. I noticed some unit tests emit them; let's change log to ERROR separately after better understanding it.